[WIP] Virtual attributes for Containers and friends (performance changes)

kbrock opened the pull request:
depends upon:

- [x] https://github.com/ManageIQ/activerecord-virtual_attributes/pull/120
- ~~https://github.com/ManageIQ/activerecord-virtual_attributes/pull/127~~

@agrare Can we outright claim that `PersistentVolume#parent` is always an `Ems`?
I get that a `ContainerVolume#parent` is polymorphic, but curious about the child class (i.e.: `PersistentVolume`). If not, may need to drop that change.

## Numbers

|     ms |query | qry ms |     rows |` comments`
|    ---:|  ---:|   ---:|      ---:|  ---
|  347.2 |   32 |  53.2 |      128 |` /container_node/report_data#node-index`
|  308.8 |   11 |  10.3 |       28 |` /container_node/report_data#node-after`
|  11.1% |  66% | 80.6% |      78% | diff

|     ms |query | qry ms |     rows |` comments`
|    ---:|  ---:|   ---:|      ---:|  ---
|  428.0 |   32 | 123.2 |       88 |` /container_group/report_data#group-index`
|  336.5 |   11 |  65.5 |       28 |` /container_group/report_data#group-after`
|  21.4% |  66% | 46.8% |      68% | diff

|     ms |query | qry ms |     rows |` comments`
|    ---:|  ---:|   ---:|      ---:|  ---
|  424.3 |   51 |  98.7 |       48 |` /container_image/report_data`
|  345.1 |   11 |  34.4 |       28 |` /container_image/report_data`
|   18.7%|   78%|  65.1%|       42%| diff

---

no longer:

|     ms |query | qry ms |     rows |` comments`
|    ---:|  ---:|   ---:|      ---:|  ---
|  278.6 |   31 |  12.7 |       48 |` /persistent_volume/report_data#before`
|  258.7 |   11 |   9.9 |       28 |` /persistent_volume/report_data#after`
|   7.1% |   65%|  22.0%|      42% | diff

kbrock committed:
add ContainerImageRegistry#full_name arel

add ContainerImage#display_registry delegate to the full_name

|     ms |query | qry ms |     rows |` comments`
|    ---:|  ---:|   ---:|      ---:|  ---
|  424.3 |   51 |  98.7 |       48 |` /container_image/report_data`
|  345.1 |   11 |  34.4 |       28 |` /container_image/report_data`
|   18.7%|   78%|  65.1%|       42%| diff

Localization of "Unknown image source" has been removed and moved
into the UI layer

kbrock commented:
update:
- rebased master
- updated test to mark now sql friendly virtual attribute as ... sql friendly.

kbrock commented:
update:
- extracted container volume assumption that the persistent volume is always associated with an ems.

The savings are very good, but I am not ready to put my foot down on that one.
The other changes are pretty much a no brainer.

update:
- rubocops: add :dependent, :inverse_of

kbrock commented:
WIP: confusing CI values

I'm very confused by the failure. This works locally and I can't imagine where this could be going wrong. All test failures are the same. (I wrote the specs at the same time)


The spec(s) only shows partial SQL, but it looks like `where(:id => subject.id)` is sending the `container_node.id` as text instead of a numeric.

```
# rspec ./spec/models/container_group_spec.rb:53
# rspec ./spec/models/container_group_spec.rb:61
# rspec ./spec/models/container_node_spec.rb:50 <==

PG::UndefinedFunction: ERROR:  operator does not exist: bigint = text
LINE 1: ...OM "container_nodes" WHERE "container_nodes"."id" = $1 ORDER...
```

Again, the sql generated locally is working fine.

Also interesting:

```ruby
subj = described_class.where(:id => container_group.id)
expect(subj.first.ready_condition_status).to eq("None") # works
expect(subj.select(:ready_condition_status).first.ready_condition_status).to eq("None") # failure
```

Fryguy commented on app/models/manageiq/providers/inventory/publisher.rb:
I think this file was accidentally committed to this PR.

kbrock committed:
add ContainerGroup#ready_condition_status

similar to 773fea35b9

already introduced indexes, so this is just ruby changes

|     ms |query | qry ms |     rows |` comments`
|    ---:|  ---:|   ---:|      ---:|  ---
|  347.2 |   32 |  53.2 |      128 |` /container_node/report_data#node-index`
|  308.8 |   11 |  10.3 |       28 |` /container_node/report_data#node-after`
|  11.1% |  66% | 80.6% |      78% | diff

|     ms |query | qry ms |     rows |` comments`
|    ---:|  ---:|   ---:|      ---:|  ---
|  428.0 |   32 | 123.2 |       88 |` /container_group/report_data#group-index`
|  336.5 |   11 |  65.5 |       28 |` /container_group/report_data#group-after`
|  21.4% |  66% | 46.8% |      68% | diff

kbrock commented:
update:
- rebase
- fixed extra file committed (thnx fry)

Fryguy commented on app/models/container_image.rb:
```diff
@@ -46,7 +46,7 @@ class ContainerImage < ApplicationRecord
   serialize :environment_variables, Hash
 
-  virtual_column :display_registry, :type => :string
+  virtual_delegate :display_registry, :to => "container_image_registry.full_name", :allow_nil => true, :type => :string
```
What happens when this returns nil?  Is there a presentation-side change where it will say "Unknown image source", or does it leave it blank?

Fryguy commented on spec/models/container_image_registry_spec.rb:
```diff
@@ -1,9 +1,19 @@
-    reg = ContainerImageRegistry.new(:name => "docker.io", :host => "docker.io")
-    expect(reg.full_name).to eq("docker.io")
+  describe "#full_name" do
+    it "works with no port" do
```
Can you also add a spec where the port is `""` since there's a case for that?

Fryguy commented:
Overall LGTM - one real question, and some specs are failing, but maybe they go away with a rebase since Ruby 2.7 is gone now.
